fix(javascript): ternary with `--use-tabs`

ikatyang opened the pull request:
Rewrite #3277, fixes #2771.

Rules:
- trailing alignments:
  - still spaces.
- ~~middle alignments:~~ Removed, see https://github.com/prettier/prettier/pull/3745#issuecomment-357953038.
  - ~~every `tabsWidth` spaces will be transformed into a tab.~~

Example:
- before
  - `tabWidth = 2`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><tab>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><tab>`
  - `tabWidth = 4`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><tab>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><tab>`
- after
  - `tabWidth = 2`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><2 space>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><2 space>`
  - `tabWidth = 4`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><2 space>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><2 space>`

P.S. `align` with string `n` is not affected.

Diff:

`{ useTabs: true, tabWidth: 2 | 4 }`
![image](https://user-images.githubusercontent.com/8341033/35029315-5888ecd0-fb95-11e7-9b86-be85a3418d44.png)

---

**Prettier pr-3745**
[Playground link](https://deploy-preview-3745--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBDAOlABDrB+LYTXErVJLABmNIF8acKjtSdzCHWT3qWud6ffvg5DhuHp3GDxjUbNIBPJlIUSKvNbhlbSBZruGSxunYbkHzrY1YGrWZ7SAA0ICAAcYAS2gBnZKDuAE5esADqXgAmMAAWyAAclK4wqABGEdFxSAAsrr6hAOYANnAAigCuEPDIAGaoRb5wyUGoXkWFAMIQALbdqMggUNBwLiCpLWAA1nAwAMruqGCFyDBB5U0gAFa+AB4AQhPTc6jdcAAyoSNIdQ0bC0GNQQOpaYpF0KON3V4raxvljQAKml-EhVutXKFHjAAAotAp9Wr1RquIJwACO5S8aLhqAR-WuyLuQQgjTCLXcA2CcEeADcRq5UEESQB3GFMhCgtC0iBRUaRCBgJG3Rm+GDClFucowdzSgBMErgtFoQA)
```sh
--tab-width 4
--use-tabs
```

**Input:**
```jsx
a
    ? {
        a: 0
      }
    : {
        a: {
             a: 0
           }
            ? {
                a: 0
              }
            : {
                y: {
                    a: 0
                }
                    ? {
                        a: 0
                    }
                    : {
                        a: 0
                    }
            }
      }
```

**Output:**
```jsx
a
	? {
			a: 0
	  }
	: {
			a: {
				a: 0
			}
				? {
						a: 0
				  }
				: {
						y: {
							a: 0
						}
							? {
									a: 0
							  }
							: {
									a: 0
							  }
				  }
	  };

```

---

<details>
<summary>old</summary>

- before
  - `tabWidth = 2`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><tab>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><tab>`
  - `tabWidth = 4`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><tab>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><tab>`
- after
  - `tabWidth = 2`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><tab><2 space>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><tab><2 space>`
  - `tabWidth = 4`
    - `<tab><2 space><tab><2 space>` -> `<tab><tab><2 space>`
    - `<tab><4 space><tab><2 space>` -> `<tab><tab><tab><2 space>`

Diff:

`{ useTabs: true, tabWidth: 2 }`
![image](https://user-images.githubusercontent.com/8341033/34925528-d3d18134-f9e4-11e7-9906-3e0500e36314.png)

`{ useTabs: true, tabWidth: 4 }`
![image](https://user-images.githubusercontent.com/8341033/34925538-e18d7eb8-f9e4-11e7-906c-4c7d5ba8d015.png)

---

**Prettier pr-3745**
[Playground link](https://deploy-preview-3745--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBDAOlABDrB+LYTXErVJLABmNIF8acKjtSdzCHWT3qWud6ffvg5DhuHp3GDxjUbNIBPJlIUSKvNbhlbSBZruGSxunYbkHzrY1YGrWZ7SAA0ICAAcYAS2gBnZKCoAE5BEADuAArBCP4oqABuEF4AJi5ovjDIAGaoADa+cK4ARkGoYADWcDAAyu5lXlAA5sgwQQCuhSDJEGDZeQWuDQVBMBGljQC2qH35nQBWvgAeAEKlFVXVqBNwADINcDMDbm0w7icATIeddUHDyCBFqEWKudBp7kENMADqKTAAFsgAByUVwfCAFb6ldz3D5wYbxA6uIJwACObS8KLGqEm0yQOVmrgKEy8LXanV8DUauTgAEU2hB4FdXDAnr9kgDkAAWFmlLy5KkAYQgEym9yg0CRIDaBQAKk9Yq0OrRaEA)
```sh
--tab-width 4
--use-tabs
```

**Input:**
```jsx
a
    ? {
        a: 0
      }
    : {
        a: {
             a: 0
           }
            ? {
                a: 0
              }
            : {
                y: {
                    a: 0
                }
                    ? {
                        a: 0
                    }
                    : {
                        a: 0
                    }
            }
      }
```

**Output:**
```jsx
a
	? {
		a: 0
	  }
	: {
		a: {
			a: 0
		}
			? {
				a: 0
			  }
			: {
				y: {
					a: 0
				}
					? {
						a: 0
					  }
					: {
						a: 0
					  }
			  }
	  };

```

</details>

ikatyang commented:
Ref: previous discussion https://github.com/prettier/prettier/pull/3277#issuecomment-344665329

> We have to read through #1026 (the PR that added tabs support) again, because as far as I remember all of this was discussed back then.
> 
> When using spaces, Prettier indents with 2 spaces + 1 indent level. When using tabs, there are three ways to go:
> 
> - Indent with one tab.
> - Indent with two tabs.
> - Indent with 2 spaces + 1 tab.
> 
> As far as I remember, all of those were discussed with pros and cons, but I don't have the time right now to dig into it.

ikatyang commented on tests/ternaries/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -222,21 +318,69 @@ aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+	  ? aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+	    ? aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+	    : aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+	  : aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
We need to use `dedentToRoot` from #3676 to re-indent with proper spaces/tab.

rhengles commented:
@ikatyang I'm amazed at the effort you put into this. I recognize that the output indeed looks very nice.

However, the reason the code is the way it is is because <tabs> _do not have an intrinsic length._

It looks very nice when you're formatting the code in your machine, with your editor tab length.

But the purpose of tabs is to be **tab length agnostic.**

When you share your tab indented code with another user who have an different tab length, the code will no longer be perfectly aligned.

That's why it is better to indent only with tabs, and to let the code be a little misaligned.

ikatyang commented:
It seems the trailing part is fine but the middle part should be still transformed into tabs per `align`?

rhengles commented:
@ikatyang Ah, the trailing part. It's simple when all you have inside a ternary is an literal, variable or object with simple properties (again, literal and variables).

But it gets messy _very_ fast - inside the object might be another object, an array or even a function, with endless combinations.

That's why I think in `useTabs` mode, every indentation level must be a `<tab>`.

ikatyang commented:
Removed the middle part and fixed the nested ternary case, @rhengles I think it should be fine now. If not, could you provide some examples so that I can improve it?

---

**Prettier pr-3745**
[Playground link](https://deploy-preview-3745--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBDdHNez3f9YA6UA-MQRZVfsWVNQ4xbeU2++sUqx79Vzz5CaUbvWETcxafWIACBXJJzg8xetRI5ABjXqAvnrlbVUdRpNHzFnVeuGz188tNO3czbcfvFDnwstvfwUAT0Dgt09dIOC-COsXO3ibaOTzOLSAlSTMqJzY-N87fRAAGhAIAAcYAEtoAGdkUFQAJxaIAHcABVaERpRUADcIGoATMrR6mGQAM1QAG3q4coAjFtQwAGs4GABlSo2aqABzZBgWgFdlkFGIMFmFpfKjpZaYLvXjgFtUB8XrgBW9QAHgAhdZbHa7VBfOAAGSOcD+TwqFxglTRACZkdcDi1XsgQCtUCsQvNoBNKi0jjAAOpjGAAC2QAA5tOUqRAlrT1pVCVS4K9Bkjyi04ABHC41MUfVDfX5IOb-cpLL41M6Xa71I7HeZwACKFwg8Bx5RgJPpoyZyAALGb1jV5jqAMIQL4-QlQaAikAXJYAFRJ-XOV30+iAA)
```sh
--tab-width 4
--use-tabs
```

**Input:**
```jsx
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
?
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
?
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
?
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa


a
    ? {
        a: 0
      }
    : {
        a: {
             a: 0
           }
            ? {
                a: 0
              }
            : {
                y: {
                    a: 0
                }
                    ? {
                        a: 0
                    }
                    : {
                        a: 0
                    }
            }
      }
```

**Output:**
```jsx
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
	? aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
		? aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
			? aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
			: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
		: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
	: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;

a
	? {
			a: 0
	  }
	: {
			a: {
				a: 0
			}
				? {
						a: 0
				  }
				: {
						y: {
							a: 0
						}
							? {
									a: 0
							  }
							: {
									a: 0
							  }
				  }
	  };

```

rhengles commented:
@ikatyang Ah, I see it now. You only ever put spaces on the _closing_ bracket of an object who is a direct child of the result part of a ternary expression.

Just so we have a comprehensive test case to ensure no regressions, can you add this test case?
Of course no one should ever write code like this, but we have to make sure nothing breaks in the future 😃 

```javascript
a
	? {
			a: function() {
				return a
					? {
							a: [
								a
									? {
											a: 0,
											b: [
												a
													? [
															0,
															1
													  ]
													: []
											]
									  }
									: [
											[
												0,
												{
													a: 0
												},
												a
													? 0
													: 1
											],
											function() {
												return a
													? {
															a: 0
													  }
													: [
															{
																a: 0
															},
															{}
													  ];
											}
									  ]
							]
					  }
					: [
							a
								? function() {
										a
											? a(
													a
														? {
																a: a(
																	{
																		a: 0
																	}
																)
														  }
														: [
																0,
																a(),
																a(
																	a(),
																	{
																		a: 0
																	},
																	a
																		? a()
																		: a(
																				{
																					a: 0
																				}
																		  )
																),
																a()
																	? {
																			a: a(),
																			b: []
																	  }
																	: {}
														  ]
											  );
										a(
											a()
												? {
														a: 0
												  }
												: (function(a) {
														return a()
															? [
																	{
																		a: 0,
																		b: a()
																	}
															  ]
															: a(
																	[
																		a
																			? {
																					a: 0
																			  }
																			: {},
																		{
																			a: 0
																		}
																	]
															  );
												  })(
														a
															? function(a) {
																	return function() {
																		return 0;
																	};
															  }
															: function(a) {
																	return function() {
																		return 1;
																	}
															  }
												  )
										);
								  }
								: function() {
										
								  }
					  ];
			}
	  }
	: a;
```

wmertens commented:
At first I was like 🤔 and then I was like 😕 but it kinda looks ok so I'll just kill that eslint mixed spaces rule and be like 🤷‍♂️

jseminck commented:
I appreciate all the effort that has gone into this but wanted to give another view. We recently ran into this issue when prettier got upgraded and are not so excited about the change:

* Our company requires all projects to use tabs
* Our company provides a set of eslint rules to projects, that are not supposed to be overridden
* Prettier can be used by teams/projects if they want to, but this is not mandatory (as it can be too opinionated for some teams, their loss...)

This change doesn't play well with this setup. We cannot disable the ESLint rule related to mixing spaces and tabs (well we can, but we are asked not to).

So for us, that means we would have to stick to the older version of prettier. 😞

duailibe commented:
@jseminck Is/are the rule/rules auto-fixable?

In that case you can use `prettier-eslint` to fix the stuff after Prettier formats everything

jseminck commented:
I am not sure if this is fix-able ( https://eslint.org/docs/rules/no-mixed-spaces-and-tabs ). I will check how that could work. Thanks for the suggestion. 👍

lydell commented:
> - Our company provides a set of eslint rules to projects, that are not supposed to be overridden
> - Prettier can be used by teams/projects if they want to

That sounds like a constant uphill battle to me. Why not allow people to use [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier)?

jseminck commented:
It hasn't caused any problems until now. We have over 100 repositories and having a shared eslint config helps switching from one project to another a bit easier. 😄 

Is there any place where I can follow how this decision was made? I just don't understand why spaces are used when the option is set to use tabs.  Especially for people who have tabs set to something else than 4 spaces I think the new approach will look a bit odd (not sure why someone would to that though, maybe some people use 2 spaces) ?

This is the difference that we noticed:

previous:
```
function test() {
	const isNotLinkedToAnyOrganisation = false;
	const api = {};
	const organisation = {};

	return isNotLinkedToAnyOrganisation
		? Promise.resolve({ data: { id: null } })
		: api.get(`/endpoint`, {
				method: 'POST',
				body: JSON.stringify({ name: organisation.org_name }),
			});
}
```

to current:
```
function test() {
	const isNotLinkedToAnyOrganisation = false;
	const api = {};
	const organisation = {};

	return isNotLinkedToAnyOrganisation
		? Promise.resolve({ data: { id: null } })
		: api.get(`/endpoint`, {
				method: 'POST',
				body: JSON.stringify({ name: organisation.org_name }),
		  });
}
```

ikatyang commented:
As you can see in your example:

```diff
function test() {
	const isNotLinkedToAnyOrganisation = false;
	const api = {};
	const organisation = {};

	return isNotLinkedToAnyOrganisation
		? Promise.resolve({ data: { id: null } })
		: api.get(`/endpoint`, {
				method: 'POST',
				body: JSON.stringify({ name: organisation.org_name }),
-			});
+		  });
}
```

(I'm not sure if everyone's browser display tabs as 4 char width, so here's the screenshot.)

![image](https://user-images.githubusercontent.com/8341033/37557786-c699fe3e-2a44-11e8-86de-ee46113946ec.png)

Its alignment looks odd before this PR if your tabs are not displayed as 2 char width, and it's not consistent with space-only:

```js
function test() {
    const isNotLinkedToAnyOrganisation = false;
    const api = {};
    const organisation = {};

    return isNotLinkedToAnyOrganisation
        ? Promise.resolve({ data: { id: null } })
        : api.get(`/endpoint`, {
              method: "POST",
              body: JSON.stringify({ name: organisation.org_name })
          });
}
```

The only solution I can think of is to add alignment spaces for it, don't know if there's a better one.

(I saw there's a [`smart-tabs`](https://eslint.org/docs/rules/no-mixed-spaces-and-tabs#options) option in ESLint, it should work in this case.)

wmertens commented:
note that the indentation after ?, : and the end is all two spaces, since ?
and : make up one space each, and thus the final indentation is always
correct, unlike when you are using tabs only. So just turn off the mixed
spaces rule and embrace prettier :)

On Sat, Mar 17, 2018, 5:41 PM Ika, <notifications@github.com> wrote:

> As you can see in your example:
>
> function test() {
> 	const isNotLinkedToAnyOrganisation = false;
> 	const api = {};
> 	const organisation = {};
>
> 	return isNotLinkedToAnyOrganisation
> 		? Promise.resolve({ data: { id: null } })
> 		: api.get(`/endpoint`, {
> 				method: 'POST',
> 				body: JSON.stringify({ name: organisation.org_name }),
>
> -			});+		  });
> }
>
> (I'm not sure if everyone's browser display tabs as 4 char width, so
> here's the screenshot.)
>
> [image: image]
> <https://user-images.githubusercontent.com/8341033/37557786-c699fe3e-2a44-11e8-86de-ee46113946ec.png>
>
> Its alignment looks odd before this PR if your tabs are not displayed as 2
> char width, and it's not consistent with space-only:
>
> function test() {
>     const isNotLinkedToAnyOrganisation = false;
>     const api = {};
>     const organisation = {};
>
>     return isNotLinkedToAnyOrganisation
>         ? Promise.resolve({ data: { id: null } })
>         : api.get(`/endpoint`, {
>               method: "POST",
>               body: JSON.stringify({ name: organisation.org_name
>
>  })
>           });
> }
>
> The only solution I can think of is to add alignment spaces for it, don't
> know if there's a better one.
>
> (I saw there's a smart-tabs
> <https://eslint.org/docs/rules/no-mixed-spaces-and-tabs#options> option
> in ESLint, it should work in this case.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/prettier/prettier/pull/3745#issuecomment-373934165>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADWlgzhnPwPxK9_DfYS9ghi0YRpMHeKks5tfTy8gaJpZM4Rd4w_>
> .
>
